default/python/AI/PolicyAI.py

```diff
@@ -0,0 +1,242 @@
+                self._adopt(infra1, True)
+
+    def _check_deadopt_centralization(self):
+        """Deadopt centralization after one turn to get rid of the IP cost and get
+        the slot for technocracy or industrialism."""
+        turns_adopted = self._empire.turnsPoliciesAdopted
+        if centralization in self._adopted and turns_adopted[centralization] != fo.currentTurn():
+            # simplified: always deadopt to free the slot for technocracy
+            self._deadopt(centralization)
+
+    def _techno_or_industry(self):
+        research_prio = self._aistate.get_priority(PriorityType.RESOURCE_RESEARCH)
+        production_prio = self._aistate.get_priority(PriorityType.RESOURCE_PRODUCTION)
+        may_switch_to_techno = self._can_adopt(technocracy, industrialism)
+        may_switch_to_industry = self._can_adopt(industrialism, technocracy)
+        # do not switch too often
+        if technocracy in self._adopted:
+            research_prio *= 1.25
+            may_switch_to_techno = False
+        elif industrialism in self._adopted:
+            production_prio *= 1.25
+            may_switch_to_industry = False
+        if may_switch_to_industry and research_prio < production_prio:
+            self._deadopt(technocracy)
+            self._adopt(industrialism)
+        elif may_switch_to_techno and research_prio > production_prio:
+            self._deadopt(industrialism)
+            self._adopt(technocracy)
+        debug(
+            "Prio res./ind. = %.1f/%.1f, maySwitch = %s/%s",
+            research_prio,
+            production_prio,
+            may_switch_to_techno,
+            may_switch_to_industry,
+        )
+
+    def _determine_influence_priority(self, new_production):
+        # priority shouldn't only be based on current value
+        forecast = self._ip + 3 * new_production
+        if infra1 in self._adopted and bureaucracy not in self._adopted and centralization not in self._adopted:
+            forecast -= self._num_populated + self._centralization_cost + self._bureaucracy_cost
+        if forecast < 40:
+            prio = 60 - forecast
+        else:
+            prio = 800 / forecast
+        prio *= (3 + self._num_populated + self._num_outposts / 10) / 10
+        # if not infra3 in self._adopted: prio += 10
+        debug("Setting influcence priority to %.1f", prio)
+        self._aistate.set_priority(PriorityType.RESOURCE_INFLUENCE, prio)
+
+    def _get_infl_prod(self, update=False):
+        """Get / Update IP production."""
+        if update:
+            self._universe.updateMeterEstimates(self._populated_planet_ids)
+            fo.updateResourcePools()
+        return self._empire.resourceProduction(fo.resourceType.influence)
+
+    def _can_adopt(self, name, replace_other=None):
+        policy = fo.getPolicy(name)
+        ret = name in self._empire.availablePolicies and policy.adoptionCost() <= self._ip
+        if replace_other and replace_other in self._adopted:
```

Comment:
I'll have to double-check, but I think you are right.

Reply:
Actually, you are wrong. Take the _can_adopt(industrialism, technocracy) for example. industrialism excludes technocracy, so policyPrereqsAndExclusionsOK would return False, that's why I do not call it her. Note that I do call it if replace_other is None.

The only safe way to implement _can_adopt with replace_other at the moment would be to temporarily deadopt replace_other, call policyPrereqsAndExclusionsOK, than readopt it. I do not like that, not the least because it would create lots of pseudo orders. Perhaps we can add replace_other as an argument to policyPrereqsAndExclusionsOK. For the moment the function works good enough, I'll add a comment.